---
title: " `r params$ids` Summary stats and quality control results"
subtitle: '`r params$pair_ids`'
author: "Asli Munzur"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide

params:
  pair_ids: 'pair_ids'
---

In this document, we show the results of quality control on the samples. The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features.  

While filtering low quality cells, we consider the following:  
- Total number of RNA reads detected in each cell  
- The number of unique genes detected in each cell  
Low-quality cells or empty droplets will often have very few genes while cell doublets or multiplets may exhibit an aberrantly high gene count.  
- The percentage of reads that map to the mitochondrial genome  
Low-quality / dying cells often exhibit extensive mitochondrial contamination. We restrict the mitochondrial content to 25% of all the reads.  
- The percentage of reads that ribosomal RNA  
Ribosomal content is restricted to 60% per cell.  


```{css, echo=FALSE}
    body .main-container {
      max-width: 90% !important;
      width: 90% !important;
    }

    body {
      max-width: 90% !important;
    }
```

```{r include = FALSE}
# source the files for the functions we need 
library(here)
source(here('pipeline', 'sourceFiles', 'utilities.R'))

```

```{r include = FALSE}

#extract the name_list from the parameters. we use the name list later on while making graphs. 
ids <- strsplit(params$pair_ids, "-")[[1]] # split by "-" SO ID must NOT CONTAIN "-"
dh_organoid_ids <- list(ids[[1]], ids[[2]])

```

```{r include = FALSE}

# extract some information 
sce_raws <- lapply(dh_organoid_ids, function(id) readRDS(here('..', 'data', 'processed', id, 'sce.rds')))
sce_qcs <- lapply(dh_organoid_ids, function(id) readRDS(here('..', 'data', 'qc', id, 'sce_qc.rds')))

# combine the sces 
sce_raws_combined <- do.call(combine_sces, sce_raws)
sce_qcs_combined <- do.call(combine_sces, sce_qcs)

seurat_raws <- lapply(sce_raws, function(sce) CreateSeuratObject(counts = counts(sce), project = sce$id))
seurat_qc <- lapply(sce_qcs, function(sce) CreateSeuratObject(counts = counts(sce), project = sce$id))

df_raws <- lapply(sce_raws, function(sce) colData(sce)) # this one has metrics before removing low quality cells 
df_qcs <- lapply(sce_qcs, function(sce) colData(sce)) # this one has metrics after removing loq quality cells 

```

```{r include = FALSE}
# here we write a function to help us make dataframes that good for making plots later on 

makeTable <- function(df, index) {
  
  # first we must go from S4 to data frame to be able to use dplyr
  df <- as.data.frame(df) %>% 
    
    # get the mean for each column we are interested in 
    summarize(sum = as.integer(mean(sum)), 
              detected = as.integer(mean(detected)), 
              subsets_mito_sum = as.integer(mean(subsets_mito_sum)), 
              subsets_mito_percent = as.integer(mean(subsets_mito_percent)), 
              subsets_ribo_sum = as.integer(mean(subsets_ribo_sum)), 
              subsets_ribo_percent = as.integer(mean(subsets_ribo_percent))) %>% 

    mutate(id = dh_organoid_ids[[index]]) %>% 

    # add a new column with the number of cells 
    mutate(cell_number = nrow(df)) %>% 
    
    # here we take advantage of the select function to reorder the columns 
    dplyr::select(id, 
                  cell_number, 
                  sum, 
                  detected, 
                  subsets_mito_sum, 
                  subsets_mito_percent, 
                  subsets_ribo_sum, 
                  subsets_ribo_percent) %>% 
    
    # and here we just rename them for clarity 
    rename(RNA_reads = sum, 
           detected_genes = detected, 
           mito_reads = subsets_mito_sum, 
           mito_percent = subsets_mito_percent, 
           ribo_reads = subsets_ribo_sum, 
           ribo_percent = subsets_ribo_percent)
  
  return(df)
  
} # end of function 

makeTableMedian <- function(df, index) {
  
  # first we must go from S4 to data frame to be able to use dplyr
  df <- as.data.frame(df) %>% 
    
    mutate(RNA_reads = as.integer(median(sum)), 
           detected_genes = as.integer(median(detected)), 
           mito_reads = as.integer(median(subsets_mito_sum)), 
           mito_percent = as.integer(median(subsets_mito_percent)), 
           ribo_reads = as.integer(median(subsets_ribo_sum)), 
           ribo_percent = as.integer(median(subsets_ribo_percent))) %>% 
    
    summarize(RNA_reads = as.integer(mean(RNA_reads)), 
              detected_genes = as.integer(mean(detected_genes)), 
              mito_reads = as.integer(mean(mito_reads)), 
              mito_percent = as.integer(mean(mito_percent)), 
              ribo_reads = as.integer(mean(ribo_reads)), 
              ribo_percent = as.integer(mean(ribo_percent))) %>% 
    
    mutate(id = dh_organoid_ids[[index]]) %>% 
    
    dplyr::select(id, 
                  RNA_reads, 
                  detected_genes, 
                  mito_reads, 
                  mito_percent, 
                  ribo_reads, 
                  ribo_percent) 
                
  return(df)
  
} # end of function 


```

```{r echo = FALSE, warning = FALSE}
# we gotta add some tables here before we jump into graphs. 
# we first make a very very basic table, then make it look better. 

BEFOREcontrolTable <- makeTable(df_raws[[1]], 1)
BEFOREexpTable <- makeTable(df_raws[[2]], 2)

# combine data from control and treatment in one table 
BEFOREtable <- rbind(BEFOREcontrolTable, BEFOREexpTable)

AFTERcontrolTable <- makeTable(df_qcs[[1]], 1)
AFTERexpTable <- makeTable(df_qcs[[2]], 2)

# combine data from control and treatment in one table 
AFTERtable <- rbind(AFTERcontrolTable, AFTERexpTable)

# MEDIAN TABLE 
BEFOREcontrolTable_median <- makeTableMedian(df_raws[[1]], 1)
BEFOREexpTable_median <- makeTableMedian(df_raws[[2]], 2)

# combine data from control and treatment in one table 
BEFOREtable_median <- rbind(BEFOREcontrolTable_median, BEFOREexpTable_median)

AFTERcontrolTable_median <- makeTableMedian(df_qcs[[1]], 1)
AFTERexpTable_median <- makeTableMedian(df_qcs[[2]], 2)

# combine data from control and treatment in one table 
AFTERtable_median <- rbind(AFTERcontrolTable_median, AFTERexpTable_median)


```

```{r echo = FALSE, warning = FALSE, results = 'asis'}
# here we just make the tables look nice and add a caption as well. 
BEFOREtable %>% 
  knitr::kable(caption = 'Table 1: Summary statistics showing cell number and MEAN values per cell- BEFORE quality control') %>% 
  kable_styling(full_width = F)
```

```{r echo = FALSE, warning = FALSE, results = 'asis'}
AFTERtable %>% 
  knitr::kable(caption = 'Table 2: Summary statistics showing cell number and MEAN values per cell- AFTER quality control') %>% 
  kable_styling(full_width = F)
```

```{r echo = FALSE, warning = FALSE, results = 'asis'}
BEFOREtable_median %>% 
  knitr::kable(caption = 'Table 3: Summary statistics showing cell number and MEDIAN values per cell- BEFORE quality control') %>% 
  kable_styling(full_width = F)
```

```{r echo = FALSE, warning = FALSE, results = 'asis'}
AFTERtable_median %>% 
  knitr::kable(caption = 'Table 4: Summary statistics showing cell number and MEDIAN values per cell- AFTER quality control') %>% 
  kable_styling(full_width = F)
```


```{r include = FALSE}

all_raw <- merge(seurat_raws[[1]], seurat_raws[[2]], add.cell.ids = c(dh_organoid_ids[[1]], dh_organoid_ids[[2]]))
all_qc <- merge(seurat_qc[[1]], seurat_qc[[2]], add.cell.ids = c(dh_organoid_ids[[1]], dh_organoid_ids[[2]]))

all_raw[["percent_mito"]] <- PercentageFeatureSet(all_raw, pattern = "^MT-")
all_qc[["percent_mito"]] <- PercentageFeatureSet(all_qc, pattern = "^MT-")

all_raw[["percent_ribo"]] <- PercentageFeatureSet(all_raw, pattern = "^RP[SL]")
all_qc[["percent_ribo"]] <- PercentageFeatureSet(all_qc, pattern = "^RP[SL]")

feats <- c("nFeature_RNA","nCount_RNA","percent_mito","percent_ribo")

```

#### BEFORE QUALITY CONTROL
```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 13}

VlnPlot(all_raw, group.by= "orig.ident", features = feats, pt.size = 0.1,ncol = 4) 

```


#### AFTER QUALITY CONTROL
```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 13}

VlnPlot(all_qc, group.by= "orig.ident", features = feats, pt.size = 0.1,ncol = 4)

```


We can also look into the relationship between certain values in scatter plots. Pearson correlation between the two features is displayed above the plot. 

#### BEFORE QUALITY CONTROL 

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 17}

cowplot::plot_grid(ncol = 4,
  FeatureScatter(all_raw, "nCount_RNA"  , "nFeature_RNA", group.by = "orig.ident", pt.size = .5),
  FeatureScatter(all_raw, "percent_mito", "nFeature_RNA", group.by = "orig.ident", pt.size = .5),
  FeatureScatter(all_raw, "percent_ribo", "nFeature_RNA", group.by = "orig.ident", pt.size = .5),
  FeatureScatter(all_raw, "percent_ribo", "percent_mito", group.by = "orig.ident", pt.size = .5)
)

```

#### AFTER QUALITY CONTROL 

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 17}

cowplot::plot_grid(ncol = 4,
  FeatureScatter(all_qc, "nCount_RNA"  , "nFeature_RNA", group.by = "orig.ident", pt.size = .5),
  FeatureScatter(all_qc, "percent_mito", "nFeature_RNA", group.by = "orig.ident", pt.size = .5),
  FeatureScatter(all_qc, "percent_ribo", "nFeature_RNA", group.by = "orig.ident", pt.size = .5),
  FeatureScatter(all_qc, "percent_ribo", "percent_mito", group.by = "orig.ident", pt.size = .5)
)

```

### Dimensionality reduction plots 

Next we include some dimensionality reduction plots. Plots on the left are before quality control, plots on the right are after. 

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 15}

# Calculate number of cells per cluster from object@ident
cell.num <- table(all_raw@active.ident)

# Add cell number per cluster to cluster labels
ClusterLabels = paste("Cluster", names(cell.num), paste0("(n = ", cell.num, ")"))

# Order legend labels in plot in the same order as 'ClusterLabels'
ClusterBreaks = names(cell.num)

all_raw <- FindVariableFeatures(all_raw)
all_raw <- ScaleData(all_raw)

all_raw <- RunPCA(all_raw, verbose = FALSE)
all_raw <- RunUMAP(all_raw, dims = 1:30)
all_raw <- RunTSNE(all_raw, dims = 1:30, check_duplicates = FALSE)

p1 <- DimPlot(all_raw, reduction = 'pca', group.by = 'orig.ident', combine = FALSE)
# p2 <- DimPlot(all_raw, reduction = 'tsne', group.by = 'orig.ident', combine = FALSE, do.label = TRUE)
p3 <- DimPlot(all_raw, reduction = 'umap', group.by = 'orig.ident', combine = FALSE)

# Plot tSNE with new legend labels for clusters
p2 <- DimPlot(object = all_raw, 
              reduction = 'tsne', 
              do.return = T) +
  scale_colour_discrete(breaks = ClusterBreaks, 
                        labels = ClusterLabels) +
  labs(x = "t-SNE 1",
       y = "t-SNE 2")


# Calculate number of cells per cluster from object@ident
cell.num <- table(all_qc@active.ident)

# Add cell number per cluster to cluster labels
ClusterLabels = paste("Cluster", names(cell.num), paste0("(n = ", cell.num, ")"))

# Order legend labels in plot in the same order as 'ClusterLabels'
ClusterBreaks = names(cell.num)


all_qc <- FindVariableFeatures(all_qc)
all_qc <- ScaleData(all_qc)

all_qc <- RunPCA(all_qc, verbose = FALSE)
all_qc <- RunUMAP(all_qc, dims = 1:30)
all_qc <- RunTSNE(all_qc, dims = 1:30, check_duplicates = FALSE)

# Plot tSNE with new legend labels for clusters
p5 <- DimPlot(object = all_qc, 
              reduction = 'tsne', 
              do.return = T) +
  scale_colour_discrete(breaks = ClusterBreaks, 
                        labels = ClusterLabels) +
  labs(x = "t-SNE 1",
       y = "t-SNE 2")


p4 <- DimPlot(all_qc, reduction = 'pca', group.by = 'orig.ident', combine = FALSE)
# p5 <- DimPlot(all_qc, reduction = 'tsne', group.by = 'orig.ident', combine = FALSE)
p6 <- DimPlot(all_qc, reduction = 'umap', group.by = 'orig.ident', combine = FALSE)

```

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 15}

grid.arrange(ggplotify::as.grob(p1[[1]]), ggplotify::as.grob(p4[[1]]), nrow = 1)

```

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 15}

grid.arrange(p2, p5, nrow = 1)

```

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 15}

grid.arrange(ggplotify::as.grob(p3[[1]]), ggplotify::as.grob(p6[[1]]), nrow = 1)

```

### Most variable features  
We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets. Here we find the 3000 most variable features (genes) in two datasets **after** quality control. Datatables below have include these genes, but we label the top 15 genes for both datasets. 

```{r include = FALSE}

seurat_qc[[1]] <- FindVariableFeatures(seurat_qc[[1]], selection.method = "vst", nfeatures = 3000)
seurat_qc[[2]] <- FindVariableFeatures(seurat_qc[[2]], selection.method = "vst", nfeatures = 3000)


# Identify the 10 most highly variable genes
top10_1 <- head(VariableFeatures(seurat_qc[[1]]), 15)
top10_2 <- head(VariableFeatures(seurat_qc[[2]]), 15)

# all the variable genes 
one_all <- VariableFeatures(seurat_qc[[1]])
two_all <- VariableFeatures(seurat_qc[[2]])


# plot variable features with labels
plot1 <- VariableFeaturePlot(seurat_qc[[1]])
plot2 <- LabelPoints(plot = plot1, points = top10_1, repel = TRUE)
plot2 <- plot2 + ggtitle(paste(dh_organoid_ids[[1]])) + theme(legend.position = "bottom")


plot3 <- VariableFeaturePlot(seurat_qc[[2]])
plot4 <- LabelPoints(plot = plot3, points = top10_2, repel = TRUE)
plot4 <- plot4 + ggtitle(paste(dh_organoid_ids[[2]])) + theme(legend.position = "bottom")



```


```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 10}

grid.arrange(plot2, plot4, nrow = 1)

```


#### Most variable genes in `r paste(dh_organoid_ids[[1]])`  
```{r}

DT::datatable(HVFInfo(seurat_qc[[1]])[VariableFeatures(seurat_qc[[1]]), ],
  extensions = 'Buttons',
  options = list(dom = 'Bfrtip',
                 buttons = list("copy", "excel", "csv", "pdf")))


```


#### Most variable genes in `r paste(dh_organoid_ids[[2]])`  
```{r}

DT::datatable(HVFInfo(seurat_qc[[2]])[VariableFeatures(seurat_qc[[2]]), ],
  extensions = 'Buttons',
  options = list(dom = 'Bfrtip',
                 buttons = list("copy", "excel", "csv", "pdf")))


```

```{r include = FALSE}

#### Number of variable features  
# Here we find the number of variable features with various cutoffs after quality control. 

qc_sample1 <- seurat_qc[[1]]
qc_sample2 <- seurat_qc[[2]]

qc_sample1 <- NormalizeData(qc_sample1)
qc_sample2 <- NormalizeData(qc_sample2)


qc_sample1 <- FindVariableFeatures(qc_sample1, selection.method = 'dispersion', mean.cutoff = c(0.005, 1))

qc_sample2 <- FindVariableFeatures(qc_sample2, selection.method = 'dispersion', mean.cutoff = c(0.005, 1))

getNumberOfHVG <- function(object, threshold){
  
  df <- as.data.frame(HVFInfo(object = object)) %>% 
    dplyr::select(dispersion.scaled) %>% 
    dplyr::filter(dispersion.scaled > threshold)
  
  return(nrow(df))
  
}

a <- getNumberOfHVG(qc_sample1, 0.5)
b <- getNumberOfHVG(qc_sample1, 1.0)
c <- getNumberOfHVG(qc_sample1, 1.5)
d <- getNumberOfHVG(qc_sample1, 2.0)
e <- getNumberOfHVG(qc_sample1, 2.5)

f <- getNumberOfHVG(qc_sample2, 0.5)
g <- getNumberOfHVG(qc_sample2, 1.0)
h <- getNumberOfHVG(qc_sample2, 1.5)
i <- getNumberOfHVG(qc_sample2, 2.0)
j <- getNumberOfHVG(qc_sample2, 2.5)

HVG_sample1 <- c(a, b, c, d, e)
HVG_sample2 <- c(f, g, h, i, j)
dispersion_cutoff <- c(0.5, 1.0, 1.5, 2.0, 2.5)

df <- as.data.frame(cbind(dispersion_cutoff, HVG_sample1, HVG_sample2))

col1Name <- paste('HVG', dh_organoid_ids[[1]], sep = '_')
col2Name <- paste('HVG', dh_organoid_ids[[2]], sep = '_')



# rename the columns 
names(df)[names(df) == "HVG_sample1"] <- paste('HVG', dh_organoid_ids[[1]], sep = '_')
names(df)[names(df) == "HVG_sample2"] <- paste('HVG', dh_organoid_ids[[2]], sep = '_')


```

#### Number of highly variable genes (HVG)
Number of HVG for both samples are given below for various gene dispersion (variance) cut offs.  

```{r}

df %>% 
  knitr::kable(caption = 'Table 5: Number of HVG after quality control') %>% 
  kable_styling(full_width = F)

```

#### Clustering  
Clustering cells and running a DE analysis between clusters can give an idea about different cell populations that may exist in the samples and various gene expression patterns. To better understand the composition of the clusters, we will look into these factors and show them on plots:  
- mitochondrial content (percentage)  
- ribosomal content (percentage)  
- number of detected genes  
0.5 was chosen as a resolution. 

Now let's look at the chosen resolutions in more detail. We visualize the clusters in a dimensionality reduction plots (tSNE) below. 

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 15}


# qc_sample1$seurat_clusters <- as.numeric(qc_sample1$seurat_clusters) + 1

qc_sample1 <- ScaleData(qc_sample1)

qc_sample1 <- RunPCA(qc_sample1, verbose = FALSE)
qc_sample1 <- RunUMAP(qc_sample1, dims = 1:30)
qc_sample1 <- RunTSNE(qc_sample1, dims = 1:30, check_duplicates = FALSE)

qc_sample1 <- FindNeighbors(qc_sample1, dims = 1:10)
qc_sample1 <- FindClusters(qc_sample1, resolution = 0.5)

#########################

qc_sample2 <- ScaleData(qc_sample2)

qc_sample2 <- RunPCA(qc_sample2, verbose = FALSE)
qc_sample2 <- RunUMAP(qc_sample2, dims = 1:30)
qc_sample2 <- RunTSNE(qc_sample2, dims = 1:30, check_duplicates = FALSE)

qc_sample2 <- FindNeighbors(qc_sample2, dims = 1:10)
qc_sample2 <- FindClusters(qc_sample2, resolution = 0.5)


Dimplot1 <- DimPlot(qc_sample1, reduction = "tsne") + ggtitle(paste(dh_organoid_ids[[1]]))
Dimplot2 <- DimPlot(qc_sample2, reduction = "tsne") + ggtitle(paste(dh_organoid_ids[[2]]))

grid.arrange(Dimplot1, Dimplot2, nrow = 1)

```
Now we will look into the contents of these clusters and visualize these values in the violin plots below.  

```{r include = FALSE}

# now lets compute the statistics above in plots, for that we need to add the clusters information to the sce objects 

sce_sample1 <- sce_qcs[[1]]
sce_sample2 <- sce_qcs[[2]]
  
# sample1
clusters <- qc_sample1$seurat_clusters
mito_percentage <- sce_sample1$subsets_mito_percent
ribo_percentage <- sce_sample1$subsets_ribo_percent
detected_genes <- sce_sample1$detected

# make a df from them 
sample1_info_df <- as.data.frame(cbind(clusters, mito_percentage, ribo_percentage, detected_genes))

# df makes 0 turn into 1, so subtract 1 from each row of the clusters 
sample1_info_df$clusters <- sample1_info_df$clusters - 1

# now we make the plots 
plot_sample1_mito <- ggplot(sample1_info_df, aes(x = factor(clusters), y = mito_percentage, fill = factor(clusters))) + geom_violin() + geom_jitter(height = 0.1, width = 0.08) + theme_classic() + ggtitle('mitochondrial content')

plot_sample1_ribo <- ggplot(sample1_info_df, aes(x = factor(clusters), y = ribo_percentage, fill = factor(clusters))) + geom_violin() + geom_jitter(height = 0.1, width = 0.08) + theme_classic() + ggtitle('ribosomal content') + ylim(0, 40)

plot_sample1_genes <- ggplot(sample1_info_df, aes(x = factor(clusters), y = detected_genes, fill = factor(clusters))) + geom_violin() + geom_jitter(height = 0.1, width = 0.08) + theme_classic() + ggtitle('detected genes') + ylim(0, 10000)

##################################

# sample2 (sample two)
clusters <- qc_sample2$seurat_clusters
mito_percentage <- sce_sample2$subsets_mito_percent
ribo_percentage <- sce_sample2$subsets_ribo_percent
detected_genes <- sce_sample2$detected

# make a df from them 
sample2_info_df <- as.data.frame(cbind(clusters, mito_percentage, ribo_percentage, detected_genes))

# df makes 0 turn into 1, so subtract 1 from each row of the clusters 
sample2_info_df$clusters <- sample2_info_df$clusters - 1

# now we make the plots 
plot_sample2_mito <- ggplot(sample2_info_df, aes(x = factor(clusters), y = mito_percentage, fill = factor(clusters))) + geom_violin() + geom_jitter(height = 0.1, width = 0.08) + theme_classic() + ggtitle('mitochondrial content')

plot_sample2_ribo <- ggplot(sample2_info_df, aes(x = factor(clusters), y = ribo_percentage, fill = factor(clusters))) + geom_violin() + geom_jitter(height = 0.1, width = 0.08) + theme_classic() + ggtitle('ribosomal content') + ylim(0, 40)

plot_sample2_genes <- ggplot(sample2_info_df, aes(x = factor(clusters), y = detected_genes, fill = factor(clusters))) + geom_violin() + geom_jitter(height = 0.1, width = 0.08) + theme_classic() + ggtitle('detected genes') + ylim(0, 10000)


```

#### `r paste(dh_organoid_ids[[1]])`

```{r echo = FALSE, message = FALSE, warning = FALSE, fig.height = 5, fig.width = 18}

# now organize them
grid.arrange(plot_sample1_mito, plot_sample1_ribo, plot_sample1_genes, nrow = 1)


```

#### `r paste(dh_organoid_ids[[2]])`

```{r echo = FALSE, message = FALSE, warning = FALSE, fig.height = 5, fig.width = 18}

# now organize them
grid.arrange(plot_sample2_mito, plot_sample2_ribo, plot_sample2_genes, nrow = 1)


```


