---
title: 'Correcting for batch effects'
subtitle: '`r params$ids` - `r params$id_type` '
author: 'asli'
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide

params:
  ids: 'ids'
  path_to_sce_clus: 'path_to_sce_clus'
  output_uncorrected: 'output_uncorrected'
  output_corrected: 'output_corrected'
  id_type: 'id_type'
  
---

#### Batch effects can introduce unwanted variation to samples. It's important to normalize the batch effect before attempting to analyze samples together. Here we show some dimensionality reduction plots before and after batch normalization in `r params$ids`. Analysis was done using scran and batchelor packages. 


```{r include = FALSE}

library(here)
source(here('pipeline', 'sourceFiles', 'utilities.R'))

```

```{r include = FALSE, message = FALSE}

# lol <- list('/huntsman/amunzur/scRNAseq-workflow/../data/clustered/sce/DH7/sce_clus.rds', '/huntsman/amunzur/scRNAseq-workflow/../data/clustered/sce/DH24/sce_clus.rds')

# snakemake passes the input files as one string separated by a space. we need to divide them into separate lists: 
paths <- as.list(unlist(strsplit(params$path_to_sce_clus[[1]], ' '))) 

# load the sces
sces <- lapply(paths, function(some_path) readRDS(some_path)) 

```

```{r include = FALSE}

# subset to common genes 
universe <- Reduce(intersect, lapply(sces, function(sce) rownames(sce)))
sces <- lapply(sces, function(sce) sce[universe, ])

# make the sces comparable by removing sequencing depth 
sces <- do.call(multiBatchNorm, sces)

# find highly variable genes  
vars <- lapply(sces, function(sce) modelGeneVar(sce))
combined.var <- do.call(combineVar, vars)
chosen.hvgs <- combined.var$bio > 0

# combine without any correction yet
combined <- do.call(correctExperiments, c(PARAM=NoCorrectParam(), sces))
combined <- runPCA(combined, subset_row = chosen.hvgs)
combined <- runTSNE(combined, dimred = "PCA")
combined <- runUMAP(combined, dimred = "PCA")

p1 <- plotTSNE(combined, colour_by = "id")
p2 <- plotUMAP(combined, colour_by = "id")

# batch correction! 
batch_removed <- fastMNN(combined, batch = combined$id, subset.row = chosen.hvgs)

batch_removed <- runTSNE(batch_removed, dimred = "corrected")
batch_removed <- runUMAP(batch_removed, dimred = "corrected")

p3 <- plotTSNE(batch_removed, colour_by = "batch")
p4 <- plotUMAP(batch_removed, colour_by = "batch")

```

**Before** batch correction  
```{r echo = FALSE, message = FALSE, fig.height = 4, fig.width = 8}

gridExtra::grid.arrange(p1, p2, nrow = 1)

```

**After** batch correction  
```{r echo = FALSE, message = FALSE, fig.height = 4, fig.width = 8}

gridExtra::grid.arrange(p3, p4, nrow = 1)

```

```{r include = FALSE}
# save the uncorrected object
saveRDS(combined, file = params$output_uncorrected)

# save the corrected object 
saveRDS(batch_removed, file = params$output_corrected)

```
