---
title: "Differential gene expression"
subtitle: '`r params$pair_ids`'
author: "Asli Munzur"
date: '`r format(Sys.Date(), "%B %d, %Y")`'

params:
  ids: 'pair_ids'
  sce_norm: 'sce_norm'
  output_path_paired: 'output_path_paired'
  pathwayHmark: '/huntsman/mdouglas/dh_organoid/data/gene_enrichment/input/h.all.v6.2.symbols.gmt'
  pathwayGO: '/huntsman/mdouglas/dh_organoid/data/gene_enrichment/input/c5.all.v6.2.symbols.gmt'
  pathwayReact: '/huntsman/mdouglas/dh_organoid/data/gene_enrichment/input/c2.cp.reactome.v7.0.symbols.gmt'
  
output: 
  html_document:
    code_folding: hide
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 90% !important;
      width: 90% !important;
    }

    body {
      max-width: 90% !important;
    }
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include = FALSE}

# Packages, Libraries, and Functions to import
library(here)
source(here('pipeline', 'sourceFiles', 'utilities.R'))
source('/huntsman/mdouglas/dh_organoid/source_scripts/utils_dge.R') # maxwell's dge functions 

```

## Differentially expressed genes 

```{r, message = FALSE, warning = FALSE}


plot_markers <- function(ids, ntop = 100, negative_control, randomSubsets = 100){
  
  sce_list <- lapply(paths, function(some_path) readRDS(some_path))
  
  ids <- lapply(sce_list, function(sce) unique(sce$id)) # extract ids 
  sce_list <- intersect_all(sce_list) # subset to common genes 
  sce_list <- scran_batch_norm(sce_list) # batch norm with scran
  sce <- combine_sces(sce_list = sce_list, prefix_col_name = 'id', prefix_cell_name = TRUE) # combine sces 
  
  # gene filtering 
  sce <- gene_filter(sce, gene_min_counts = 100, gene_min_detection_rate = 5, reporters = NULL)
  
  if(negative_control){
    markers <- findMarkers(sce[rowData(sce)$qc_pass, ], groups = sample(sce$id), log.p = TRUE)
  }
  else{
    markers <- findMarkers(sce[rowData(sce)$qc_pass, ], groups = sce$id, log.p = TRUE)
  }
  
  markers <- markers[[ids[[1]]]] # pos logFC = up-regulated in tranduced cells
  markers_all <- as.data.frame(markers) %>% rownames_to_column('gene_symbol')
  
  if ("FDR" %in% colnames(markers_all)) {                                             # Case when p = p.value
    markers_sig <- as.data.frame(markers) %>% rownames_to_column('gene_symbol') %>% dplyr::filter(FDR <= 0.05)
    tooLow <- function(x) (ifelse(x < 1e-315, 1e-315, x))
    markers_sig <- markers_sig %>% mutate_at(c("p.value", "FDR"), tooLow)
    # Order the genes for plotting
    markers_sig <- dplyr::arrange(markers_sig, dplyr::desc(abs(markers_sig[,5])))     # order
    t_mark <- markers_sig                                                             # make copy
    t_mark$FDR <- -log10(t_mark[,4])                                                  # log transform
    tt_top <- head(t_mark, ntop)                                                      # Identify top 'n' genes
    
  } else {                                                                            # Case when p = log.p.value
    markers_sig <- as.data.frame(markers) %>% rownames_to_column('gene_symbol') %>% dplyr::filter(log.FDR < -1.6)
    # Order the genes for plotting
    markers_sig <- dplyr::arrange(markers_sig, dplyr::desc(abs(markers_sig[,5])))     # order
    t_mark <- markers_sig                                                             # make copy
    t_mark$log.FDR <- -(t_mark[,4])
    tt_top <- head(t_mark, ntop)                                                      # Identify top 'n' genes
  }
  
  p <- ggplot(t_mark, aes_string(x = names(t_mark)[5], y = names(t_mark)[4])) +
    geom_point(colour = 'gray') +
    geom_text_repel(data = tt_top, aes(label = gene_symbol), colour = 'red', size = 5) +
    geom_point(data = tt_top, aes_string(x = names(t_mark)[5], y = names(t_mark)[4]), colour = "blue") +
    theme_minimal() +
    xlab('logFC') +
    labs(title = glue('{infected}  vs. {control} top {ntop} DEGs', infected = ids[[1]],
                      control = ids[[2]], ntop = ntop),
         subtitle = 'positive logFC means up-regulated in experimental sample') + theme(plot.subtitle = element_text(hjust = 0.5))

  # Cluster/Subset DGE Analysis by repeated sampling 
  degs <- fMarkersSampling(sce, ids[1], ids[2], randomSubsets)
  commonDEGs <- assembleMarkersByP(degs)
  samplingPlot <- plotDEGs(commonDEGs, ids, ntop, params$reporter)
  
  outputs <- list(markers_all = markers_all, markers_sig = markers_sig, tt_top = tt_top, plot = p, smarkers_sig = commonDEGs, splot = samplingPlot)
  # outputs <- list(markers_all = markers_all, markers_sig = markers_sig, tt_top = tt_top, plot = p)
}

paths <- as.list(unlist(strsplit(params$sce_norm[[1]], ' '))) 
outputs <- plot_markers(paths, ntop = 50, negative_control = FALSE)

colnames(outputs$markers_all)[5] <- "logFC"


```

```{r fig.height= 10, fig.width= 12}

if (length(clusters) < 3) {
  if(is.null(output$splot)){ # if no random sampling:
    outputs$plot
  } else { # if have random sampling:
    plot_grid(outputs$plot, outputs$splot, labels = "AUTO")
  }
} else {
  writeLines('Not yet implemented')
}

```

**positive logFC = upregulated in transduced sample, genes sorted by absolute logFC**

```{r}
outputs$markers_sig %>%
  DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list("copy", "excel", "csv", "pdf")))
```

## Gene Enrichment 

**Plots show top 10 significant pathways**

```{r}

run_gene_enrichment <- function(markers, pathways_input, nperm){
  
  pathways <- fgsea::gmtPathways(pathways_input)
  marker <- markers$markers_all %>% dplyr::select(gene_symbol, logFC) %>% na.omit() %>% distinct()

  ranks <- deframe(marker)
  fgseaRes <- fgsea::fgsea(pathways = pathways, stats = ranks, nperm = nperm) %>% as_tibble() %>% dplyr::arrange(desc(NES))
  sig_pathways <- fgseaRes[fgseaRes$padj < 0.05 , c('pathway', 'pval', 'padj', 'NES')]
  
  results <- list(fgseaRes = fgseaRes, sig_pathways = sig_pathways)
  
  return(results)
  
}

make_ge_plots <- function(fgseaRes, pathway_name, pthresh = 0.05, ntop = 10, y_text_size = 10){
  
  fgseaRes <- fgseaRes[fgseaRes$padj < pthresh , ]
  fgseaRes <- top_n(fgseaRes, n = ntop, wt = abs(NES))
  fgseaRes$pathway <- as.character(lapply(fgseaRes$pathway, function(pw) str_replace(pw, pattern = glue::glue(pathway_name, "_"), "")))
  
  if(dim(fgseaRes)[1] > 0){
    
      plot <- ggplot(fgseaRes, aes(reorder(pathway, NES), NES)) +
              geom_col(aes(fill=padj < pthresh)) +
              coord_flip() +
              labs(x="Pathway", y="Normalized Enrichment Score",
              title= glue::glue(pathway_name, " pathways")) +
              theme(text = element_text(size = 15),
              axis.text.y = element_text(size = y_text_size))
    
      return(plot)
    
    } else{
    
      return(NULL)
    
    }
}

```

```{r fig.height=5, fig.width=20, message=FALSE, warning=FALSE}

set.seed(123458)
fgseaReslist <- run_gene_enrichment(markers = outputs, pathways_input = params$pathwayHmark, nperm = 10000)
make_ge_plots(fgseaRes = fgseaReslist$fgseaRes, pathway_name = "HALLMARK", pthresh = 0.05, ntop = 10)
```

```{r}
fgseaReslist$sig_pathways %>%
  DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list("copy", "excel", "csv", "pdf")))
```

```{r fig.height=5, fig.width=20, message=FALSE, warning=FALSE}
set.seed(123458)
fgseaReslist <- run_gene_enrichment(markers = outputs, pathways_input = params$pathwayGO, nperm = 10000)
make_ge_plots(fgseaRes = fgseaReslist$fgseaRes, pathway_name = "GO", pthresh = 0.05, ntop = 10)
```

```{r}
fgseaReslist$sig_pathways %>%
  DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list("copy", "excel", "csv", "pdf")))
```

```{r fig.height=5, fig.width=20, message=FALSE, warning=FALSE}
set.seed(123458)
fgseaReslist <- run_gene_enrichment(markers = outputs, pathways_input = params$pathwayReact, nperm = 10000)
make_ge_plots(fgseaRes = fgseaReslist$fgseaRes, pathway_name = "REACTOME", pthresh = 0.05, ntop = 10)
```

```{r}
fgseaReslist$sig_pathways %>%
  DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list("copy", "excel", "csv", "pdf")))
```

## Negative Control

**Should return no DEG or very few DEGs**

```{r warning = FALSE, message = FALSE}
outputs <- plot_markers(paths, ntop = 50, negative_control = TRUE)
```

```{r}
outputs$markers_sig %>%
  DT::datatable(
    extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = list("copy", "excel", "csv", "pdf")))
```

