---
title: '`r params$ids` Comparing endometrial endometrioid and ovarian endometrioid tumors'
subtitle: 
author: "aslÄ±"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide

params:
  ova_id: 'VOA11068_ENOC-DH13-DH18-DH8-DH24-VOA11229_CCOC-DH7'
  endo_id: 'DH3-DH4-DH10-DH15-DH16-DH17'
  seed: 123
---

```{r include = FALSE}

library(here)
source(here('..', 'sourceFiles', 'utilities.R'))

ova_integrated <- readRDS(here('..', 'data', 'integrated', params$ova_id, 'integrated.rds'))
endo_integrated <- readRDS(here('..', 'data', 'integrated', params$endo_id, 'integrated.rds'))

```

In report we show the results of a clustering analysis done on the endometrial and ovarian samples, separately. The dim reduction plots shown below we done on the integrated samples with the batch effects removal. 

```{r echo = FALSE, message = FALSE, fig.height = 5, fig.width = 8}

p3 <- DimPlot(ova_integrated, reduction = 'tsne', group.by = 'id', combine = FALSE) + labs(title = 'lol')
p4 <- DimPlot(endo_integrated, reduction = 'tsne', group.by = 'id', combine = FALSE) 

p3 <- ggplotify::as.grob(p3[[1]]) + ggtitle('ovary')
p4 <- ggplotify::as.grob(p4[[1]]) + ggtitle('endometrium')

gridExtra::grid.arrange(p3, p4, nrow = 1)

DefaultAssay(ova_integrated) <- "RNA"
markers <- FindAllMarkers(ova_integrated, test.use = "LR", latent.vars = "id")

```

```{r include = FALSE}

# convert to sce 
sce.ova <- SingleCellExperiment(assays = list(logcounts = GetAssayData(object = ova_integrated)))
sce.ova <- runPCA(sce.ova)
sce.ova <- runTSNE(sce.ova)
sce.ova <- runUMAP(sce.ova)

# cluster the samples, two clusters each 
clust.kmeans.ova <- kmeans(reducedDim(sce.ova, "TSNE"), centers=5)
sce.ova$cluster <- factor(clust.kmeans.ova$cluster)

```

```{r include = FALSE}
# make the counts matrix, for that extract coldata 
col_df <- as.data.frame(colData(sce.ova))

# where cluster num is 1 
idx_1 <- which(col_df$cluster == 4)
idx_else <- which(col_df$cluster != 4)

# extract the counts data for these related cells that belong to cluster 1 
logdata <- as.data.frame(logcounts(sce.ova))

logdata_1 <- logdata[, idx_1]
logdata_else <- logdata[, idx_else]

# now combine 
combined <- as.data.frame(cbind(logdata_1, logdata_else))

group <- c(rep(1, ncol(logdata_1)), rep(2, ncol(logdata_else)))

# remove neg counts 
combined <- combined[apply(combined, 1, function(combined) all(combined >= 0)), ]



```

```{r include = FALSE}

dgeObject <- DGEList(counts = combined, genes = rownames(combined), group = group)

# filter the lowly expressed genes - have at least in 0.5 cpm in at least 2 samples 
keep <- rowSums(cpm(dgeObject) > 0.5) >= 2 
dgeObject <- dgeObject[keep, ]

designMatrix <- model.matrix(~group)

# estimate the dispersion (variance) of genes 
dgeObject <- estimateDisp(dgeObject, design = designMatrix)

# fit negative binomial model to the data 
fit <- glmQLFit(dgeObject, designMatrix, robust = TRUE)

# compare group2 to 1  
qlf <- glmQLFTest(fit, coef = 2)

```

```{r gene_enrichment, include = FALSE}

# Map gene names to entrez ids for goana and kegga. we have to convert from ensembl ids to entrez ids. 
geneIDs <- ensembldb::select(EnsDb.Hsapiens.v75, keys = rownames(qlf), keytype = "SYMBOL", columns = c("SYMBOL","ENTREZID"))

# sometimes we get duplicates, remove those 
geneIDs <- geneIDs[!duplicated(geneIDs$SYMBOL), ]
geneIDs <- geneIDs[!duplicated(geneIDs$ENTREZID), ]

# lets make sure we have the same genes in both 
idx <- is.na(geneIDs$ENTREZID)
idx_keep <- which(idx == FALSE)
qlf_modified <- qlf[idx_keep, ]

# drop the NAs, these are the genes we couldn't find an entrez id for 
geneIDs <- na.omit(geneIDs)

rownames(qlf_modified) <- as.integer(geneIDs$ENTREZID)

# do some gene enrichment 
go <- goana(qlf_modified, species= "Hs") # entrez gene ids are in mapping 
keg <- kegga(qlf_modified, geneid = geneIDs$ENTREZID, species="Hs")

```

### Diagnostic Plots 
These plots are intended to give an idea about the model and whether it is a good fit to represent the data we are interested in analyzing. 

Plot cells on a two-dimensional scatterplot so that distances on the plot approximate the expression differences between the cells  
In the plot below, the distance between each pair of samples can be interpreted as the leading log-fold change between the samples for the genes that best distinguish that pair of samples. Therefore, samples shown farther apart are more different in terms of gene expression. Ideally we would expect to see the replicates closer together. However, note that decreasing the number of dimensions down to two may not represent the data as accurately. This plot is here only to give an idea about how similar the samples are. Dimensionality reduction plots are more reliable to look at. 


```{r MDS_plot, echo = FALSE, message = FALSE, fig.height = 4, fig.width = 4}

# plotMDS(dgeObject, col = c('red', 'blue'))

```


#### Plot biological coefficient of variation against gene abundance, in log2 counts per million.  
Biological coefficient of variation is the true variation between genes, and it corresponds to the dispersion parameter of the negative binomial model. To learn more about dispersion, variation and how they relate to one another, read more about the negative binomial model and its implementation in scRNA sequencing data.  
Look for these to make sure your the model is a good fit:  
**1.** High dispersion (variance) for low counts is expected. Dispersion is expected to decrease smoothly as the abundance increases.  
**2.** For humans, common dispersion (the red line) should be around 0.3 - 0.4, meaning replicates can differ by 30% or more. If the common dispersion is more, the results of this analysis may not be as reliable.  

**This plot will give you a measure of the variation between the samples.**  
```{r, echo = FALSE, message = FALSE, fig.height = 4, fig.width = 4}
# plot biological coefficient of variation against gene abundance, in ln counts per million 
plotBCV(dgeObject)

```

#### Plot the genewise quasi-likelihood dispersion against the gene abundance (in log2 counts per million).
In this script, we squeeze the dispersion values from raw counts towards the trended dispersion. (This just means we transform the data in certain ways to make the model fit.)
The plots gives an idea about how much the data was squeezed. The expected is too see squeezed dispersion values (shown in red) closer to the trended (blue) line.  

```{r echo = FALSE, message = FALSE, fig.height = 4, fig.width = 4}
# Plot the genewise quasi-likelihood dispersion against the gene abundance (in ln counts per million).
plotQLDisp(fit)

```


#### Upregulated genes 
Now we take a closer look at the **upregulated** genes. To look for interesting genes, we do the following filtering:  
We don't show genes with log fold change (LogFC) less than 0.5. 
We also eliminate genes with p values more than 0.05.  
We resticted this table to first 50 genes. 
  
To help interpret, we use the follwing ranking system: 
Genes are ranked separately for log fold change (high to low) and p values (low to high). We then add these ranks together to get **summed_ranks.**  
Genes with smaller summed_ranks have high fold changes and low p values.

```{r echo = FALSE, message = FALSE, warning = FALSE}

# here we do some manipulation to show the upregulated genes in a meaningful way 
upregulated_genes_table <- make_upreg_table(qlf)%>%
  mutate_at(vars(logFC), funs(round(., 3)))

DT::datatable(upregulated_genes_table,
  extensions = 'Buttons',
  options = list(dom = 'Bfrtip',
                 buttons = list("copy", "excel", "csv", "pdf")))

# save the table to be loaded later 
saveRDS(upregulated_genes_table, file = here('..', 'data', 'dge', 'edgeR', paste(params$ids, 'up.rds')))

```

#### Downregulated genes 
Filtering and ranking systems is the same as upregulated genes, as explained above. Genes with the highest fold changes and lowest p values have lower summed_ranks.
  
```{r echo = FALSE, message = FALSE}

# here we do some manipulation to show the upregulated genes in a meaningful way 
downregulated_genes_table <- make_downreg_table(qlf) %>% 
  mutate_at(vars(logFC), funs(round(., 3)))

DT::datatable(downregulated_genes_table, 
  extensions = 'Buttons', 
  options = list(dom = 'Bfrtip', 
                 buttons = list("copy", "excel", "csv", "pdf")))

# save the table to be loaded later 
saveRDS(upregulated_genes_table, file = here('..', 'data', 'dge', 'edgeR', paste(params$ids, 'down.rds')))

```

#### Plot the differentially expressed genes  
The blue lines indicate 1-fold up or down. The number of genes that are significantly up or down regulated are shown in the table, according to the plot. Here, we compare sample2 to sample1.  
```{r echo = FALSE, message = FALSE, fig.height = 4, fig.width = 5}
# plot the differentially expressed genes 

plotMD(qlf, main = 'DE genes', ylab = 'log2 fold change', cex = 0.40)
abline(h=c(-1,1), col="blue")

```

Now we take a look at the DE genes in a volcano plot. Note that we label the first 40 significantly up and downregulated genes in endometrial samples compared to ovarian samples.  
Fold change cut off is 1, p value cut off is 10e-5.  
```{r echo = FALSE, message = FALSE, fig.height = 9, fig.width = 11}
# we need to work on qlf a little bit so that we can use the enhancedvolcano function 

data <- qlf$table

data <- data %>% 
  dplyr::select(logFC, PValue) %>% 
  rename(log2FoldChange = logFC, pvalue = PValue)

cool_genes <- c(upregulated_genes_table[1][1: 15, ], downregulated_genes_table[1][1: 15, ])

library(EnhancedVolcano)

EnhancedVolcano(data,
  lab = rownames(data),
  x = 'log2FoldChange',
  y = 'pvalue',
  title = 'Differentially Expressed Genes',
  subtitle = 'endometrium vs ovary', 
  selectLab = cool_genes, 
  labCol = 'black',
  labFace = 'bold',
  drawConnectors = TRUE,
  widthConnectors = 0.4,
  colConnectors = 'black',
  pointSize = 0.7, 
  xlim = c(-2, 2), 
  FCcutoff = 1, 
  pCutoff = 10e-5, 
  legend=c('not significant','Log (base 2) fold-change','P value',
      'P value & Log (base 2) fold-change'), 
  legendPosition = 'right',
  legendLabSize = 8,
  legendIconSize = 5.0)


```

Number of DE genes in endometrial tumors compared to ovarian tumors: 
```{r, echo = FALSE, message = FALSE}

summary <- summary(decideTests(qlf))

summary %>% 
    knitr::kable(caption = 'DE genes') %>% 
    kable_styling(full_width = F)
```

Same table as above, but now we only show the genes with fold change above 1. 
```{r echo = FALSE, message = FALSE}

summary <- summary(decideTests(qlf_threshold))

summary %>% 
    knitr::kable(caption = paste('DE genes')) %>% 
    kable_styling(full_width = F)

```



### Gene enrichment  
#### GO Pathway analysis  
Understanding the table:  
**Ont:**  
Note that GO shows three complementary biological concepts including Biological Process (BP), Molecular Function (MF) and Cellular Component (CC).  
You can find these in the 'Ont' column.  
**N:**  
number of genes in the GO term  
**Up:**  
number of up-regulated differentially expressed genes.  
**Down:**
number of down-regulated differentially expressed genes.
**P.Up**  
p-value for the representation of GO pathway in up-regulated genes  
**P.Down**  
p-value for the representation of GO pathway in down-regulated genes  
The above is also true for the KEGG table below. 

```{r echo = FALSE, message = FALSE}
# here we show the results of the go pathways 
gotable <- topGO(go, sort="up")

# do some cleaning up
DT::datatable(gotable, 
  extensions = 'Buttons', 
  options = list(dom = 'Bfrtip', 
                 buttons = list("copy", "excel", "csv", "pdf")))
  
```

#### KEGG analysis 

```{r echo = FALSE, message = FALSE}
# here we show the results of the KEGG analysis 
keggtable <- topKEGG(keg, sort="up")

# do some cleaning up
DT::datatable(keggtable, 
  extensions = 'Buttons', 
  options = list(dom = 'Bfrtip', 
                 buttons = list("copy", "excel", "csv", "pdf")))

```






